tests/console/libssh.pm

```diff
@@ -0,0 +1,129 @@
+# notice and this notice are preserved.  This file is offered as-is,
+# without any warranty.
+
+# Package: libssh4 libssh2-1
+# Summary: libssh version 1 and version 2 test
+#  Libssh4 or libssh2-1 are used by the following softwares respectively:
+#    libssh: curl, libvirt
+#    libssh2: qemu-block-ssh, libvirt
+#  They will be used to in the following test:
+#  * Set up a docker container as one end of client/server test
+#  * Run test cases below:
+#       curl -s sftp://
+#       curl -s scp://
+#       virsh -c "qemu+libssh://"
+#       virsh -c "qemu+libssh2://"
+#       qemu-system-x86_64 -drive file=ssh://
+#  * It's difficult to clean up libvirt so set test flag to rollback always
+# Maintainer: Tony Yuan <tyuan@suse.com>
+
+package libssh;
+use base 'consoletest';
+use strict;
+use warnings;
+use testapi;
+use utils;
+use Utils::Systemd 'disable_and_stop_service';
+use version_utils qw(is_sle get_os_release);
+use registration 'add_suseconnect_product';
+
+sub create_image {
+
+    my $tag;
+    my $pkgs = "openssh curl libvirt-client sshpass";
+
+    #Get the base image TAG
+    my ($v, $sp) = get_os_release;
+
+    if (is_sle('<15')) {
+        $tag = "sles${v}sp$sp";
+    } else
+    { $tag = "sle${v}:${v}.$sp";
+    }
+
+    record_info($tag);
+
+    # Dockerfile
+    my $dockerfile = <<'EOT';
+ARG sletag
+FROM registry.suse.com/suse/$sletag
+ARG pkgs
+ENV ADDITIONAL_MODULES PackageHub
+RUN zypper --gpg-auto-import-keys ref -s
+RUN zypper install -y $pkgs
+RUN echo "root:test" | chpasswd
+RUN /usr/sbin/sshd-gen-keys-start
+CMD ["/usr/sbin/sshd", "-D"]
+EOT
+    #Build image with openssh curl libvirt-client sshpass
+    assert_script_run("mkdir /tmp/build && cd /tmp/build");
+    assert_script_run("cat > /tmp/build/Dockerfile <<'END'\n$dockerfile\nEND\n( exit \$?)");
+    assert_script_run(qq(docker build -t libssh_image --build-arg sletag=$tag --build-arg pkgs="$pkgs" .), timeout => 600);
+}
+
+sub run {
+    select_console 'root-console';
+
+    if (is_sle('<15')) {
+        add_suseconnect_product("sle-module-containers", 12);
+    } else { add_suseconnect_product("sle-module-containers");
+    }
+    add_suseconnect_product("PackageHub");
+    zypper_call("in docker libvirt-daemon-qemu qemu-kvm sshpass qemu-block-ssh");
+    systemctl("start docker.service libvirtd.service");
+    create_image;
+    assert_script_run("iptables -I INPUT 5 -p tcp --dport 22 -j ACCEPT");
+    assert_script_run("mkdir /tmp/test; echo -n libssh_testcase001 > /tmp/test/libssh_testfile");
+    assert_script_run("qemu-img create -f raw /tmp/test/libssh_block.raw 20m");    #file to be uploaded to container for qemu block test
+    assert_script_run('docker run -itd --name libssh_container --hostname container01 --add-host susetest:$(ip -4 addr show docker0 | grep -Po "inet \K[\d.]+") --privileged libssh_image');
+
+    #start container as test client
+    type_string("docker exec -it libssh_container bash\n", wait_still_screen => 3);
```

Comment:
Not sure what `wait_still_screen` does in `root-console` mode. Is this necessary?

Reply:
> Not sure what `wait_still_screen` does in `root-console` mode. Is this necessary?

All other api like "script_run" need send back the shell command result "$?" via ttyS0 but "docker exec" will take the test inside container. The container will take over the ttyS0 .  ttyS0 will not be give back to host until the test inside container is done. It's not possible to know how much time it take to give ttyS0 back to host. It likely cause timeout if script_run is used.